Amplify/Categories/DataStore/Model/Collection/List+Model.swift

```diff
@@ -7,125 +7,165 @@
+    static func decode<ModelType: Model>(decoder: Decoder,
+                                         modelType: ModelType.Type) -> List<ModelType>
+
+}
+
+/// `List<ModelType>` is a custom `Collection` for accessing `Model` elements. It provides basic JSON array decoding,
+/// `Collection`, and `ExpressibleByArrayLiteral` conformance to access an array of `Model` elements. It does not
+/// provide `Paginatable` and `LazyLoad` conformance and is implemented by the plugin specific subclass of
+/// `List<ModelType>`.
+///
+/// The decoding logic uses `ModelListRegistry` to check for decodability and vends instances of subclasses of
+/// `List<ModelType>` to instantiate itself to provide plugin specific capabilities.
+open class List<ModelType: Model>: ModelList {
+    public typealias ModelListElement = ModelType
+    public typealias Page = List<ModelType>
+    public typealias Element = ModelType
+    public typealias Elements = [Element]
+
+    /// The array of `Element` that backs the custom collection implementation.
+    internal var elements: Elements
+
+    // MARK: - Initializers
 
-    // MARK: - ExpressibleByArrayLiteral
+    public init(_ elements: Elements) {
+        self.elements = elements
+    }
 
-    required convenience public init(arrayLiteral elements: Element...) {
+    public required convenience init(arrayLiteral elements: ModelType...) {
         self.init(elements)
     }
 
     // MARK: - Collection conformance
 
-    public var startIndex: Index {
-        loadIfNeeded()
-        return elements.startIndex
+    open var startIndex: Int {
+        elements.startIndex
     }
 
-    public var endIndex: Index {
-        return elements.endIndex
+    open var endIndex: Int {
+        elements.endIndex
     }
 
-    public func index(after index: Index) -> Index {
-        return elements.index(after: index)
+    open func index(after index: Index) -> Index {
+        elements.index(after: index)
     }
 
-    public subscript(position: Int) -> Element {
-        return elements[position]
+    open subscript(position: Int) -> Element {
+        elements[position]
     }
 
-    public __consuming func makeIterator() -> IndexingIterator<Elements> {
-        loadIfNeeded()
-        return elements.makeIterator()
+    open __consuming func makeIterator() -> IndexingIterator<Elements> {
+        elements.makeIterator()
     }
 
-    // MARK: - Persistent Operations
+    // MARK: Persistant operations
 
+    @available(*, deprecated, message: "Not supported.")
     public var totalCount: Int {
         // TODO handle total count
         return 0
     }
 
-    public func limit(_ limit: Int) -> Self {
-        // TODO handle query with limit
-        self.limit = limit
-        state = .pending
-        return self
+    // MARK: Paginatable
+
+    open func hasNextPage() -> Bool {
+        return false
+    }
+
+    open func getNextPage(limit: Int? = nil, onComplete: @escaping PageResult) {
```

Comment:
I played around with something like this on Android, but ultimately we implemented a `getRequestForNextResult` instead, which returns a `GraphQLRequest`, rather than actually executing it.   I think this is a better approach because:
 1.  Better DX - The logic for making subsequent requests is congruent with the first request (Call `Amplify.API.query` for each one)
 2. More extensible - Customers may want the option to modify the `GraphQLRequest` prior to executing it (perhaps apply a decorator to modify some other variable).  By providing the `GraphQLRequest` to the customer, it keeps this possibility available.

Reply:
thanks for calling this out, i'll review with the iOS folks on these two options and call out that this is what's on Android https://github.com/aws-amplify/amplify-android/blob/main/core/src/main/java/com/amplifyframework/api/graphql/PaginatedResult.java